### ISSUE TITLE
Cannot download from private repository on Hugging Face using optimum 1.17

### ISSUE THREAD

**Issue by idObTri** [huggingface/optimum#1737]
### System Info

```shell
optimum 1.17, (any variant of OS; I tested on macOS 14, Ubuntu 22.04, Windows 10), Python 3.10 or 3.11 based on miniforge3

With previous version of optimum (<1.17), this code works properly.
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

model = ORTModelForCustomTasks.from_pretrained([model_name], use_auth_token = [hf_token])

### Expected behavior

---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/huggingface_hub/utils/_errors.py:304, in hf_raise_for_status(response, endpoint_name)
    303 try:
--> 304     response.raise_for_status()
    305 except HTTPError as e:

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/requests/models.py:1021, in Response.raise_for_status(self)
   1020 if http_error_msg:
-> 1021     raise HTTPError(http_error_msg, response=self)

HTTPError: 401 Client Error: Unauthorized for url: https://huggingface.co/api/models/[username]/[repository]/tree/main?recursive=True&expand=False

The above exception was the direct cause of the following exception:

RepositoryNotFoundError                   Traceback (most recent call last)
Cell In[5], line 1
----> 1 model = ORTModelForCustomTasks.from_pretrained(model_name, use_auth_token = hf_token)

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/optimum/onnxruntime/modeling_ort.py:662, in ORTModel.from_pretrained(cls, model_id, export, force_download, use_auth_token, cache_dir, subfolder, config, local_files_only, provider, session_options, provider_options, use_io_binding, **kwargs)
    613 @classmethod
    614 @add_start_docstrings(FROM_PRETRAINED_START_DOCSTRING)
    615 def from_pretrained(
   (...)
    629     **kwargs,
    630 ):
    631     """
    632     provider (`str`, defaults to `"CPUExecutionProvider"`):
    633         ONNX Runtime provider to use for loading the model. See https://onnxruntime.ai/docs/execution-providers/ for
   (...)
    660         `ORTModel`: The loaded ORTModel model.
    661     """
--> 662     return super().from_pretrained(
    663         model_id,
    664         export=export,
    665         force_download=force_download,
    666         use_auth_token=use_auth_token,
    667         cache_dir=cache_dir,
    668         subfolder=subfolder,
    669         config=config,
    670         local_files_only=local_files_only,
    671         provider=provider,
    672         session_options=session_options,
    673         provider_options=provider_options,
    674         use_io_binding=use_io_binding,
    675         **kwargs,
    676     )

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/optimum/modeling_base.py:349, in OptimizedModel.from_pretrained(cls, model_id, export, force_download, use_auth_token, cache_dir, subfolder, config, local_files_only, trust_remote_code, revision, **kwargs)
    344         logger.warning(
    345             f"The argument `revision` was set to {revision} but will be ignored for {model_id.split('@')[1]}"
    346         )
    347     model_id, revision = model_id.split("@")
--> 349 library_name = TasksManager.infer_library_from_model(model_id, subfolder, revision, cache_dir)
    351 if library_name == "timm":
    352     config = PretrainedConfig.from_pretrained(model_id, subfolder, revision)

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/optimum/exporters/tasks.py:1654, in TasksManager.infer_library_from_model(cls, model_name_or_path, subfolder, revision, cache_dir, library_name)
   1651 if library_name is not None:
   1652     return library_name
-> 1654 all_files, _ = TasksManager.get_model_files(model_name_or_path, subfolder, cache_dir)
   1656 if "model_index.json" in all_files:
   1657     library_name = "diffusers"

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/optimum/exporters/tasks.py:1350, in TasksManager.get_model_files(model_name_or_path, subfolder, cache_dir)
   1348 if not isinstance(model_name_or_path, str):
   1349     model_name_or_path = str(model_name_or_path)
-> 1350 all_files = huggingface_hub.list_repo_files(model_name_or_path, repo_type="model")
   1351 if subfolder != "":
   1352     all_files = [file[len(subfolder) + 1 :] for file in all_files if file.startswith(subfolder)]

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/huggingface_hub/utils/_validators.py:118, in validate_hf_hub_args.<locals>._inner_fn(*args, **kwargs)
    115 if check_use_auth_token:
    116     kwargs = smoothly_deprecate_use_auth_token(fn_name=fn.__name__, has_token=has_token, kwargs=kwargs)
--> 118 return fn(*args, **kwargs)

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/huggingface_hub/hf_api.py:2779, in HfApi.list_repo_files(self, repo_id, revision, repo_type, token)
   2751 @validate_hf_hub_args
   2752 def list_repo_files(
   2753     self,
   (...)
   2758     token: Optional[Union[bool, str]] = None,
   2759 ) -> List[str]:
   2760     """
   2761     Get the list of files in a given repo.
   2762 
   (...)
   2777         `List[str]`: the list of files in a given repository.
   2778     """
-> 2779     return [
   2780         f.rfilename
   2781         for f in self.list_repo_tree(
   2782             repo_id=repo_id, recursive=True, revision=revision, repo_type=repo_type, token=token
   2783         )
   2784         if isinstance(f, RepoFile)
   2785     ]

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/huggingface_hub/hf_api.py:2779, in <listcomp>(.0)
   2751 @validate_hf_hub_args
   2752 def list_repo_files(
   2753     self,
   (...)
   2758     token: Optional[Union[bool, str]] = None,
   2759 ) -> List[str]:
   2760     """
   2761     Get the list of files in a given repo.
   2762 
   (...)
   2777         `List[str]`: the list of files in a given repository.
   2778     """
-> 2779     return [
   2780         f.rfilename
   2781         for f in self.list_repo_tree(
   2782             repo_id=repo_id, recursive=True, revision=revision, repo_type=repo_type, token=token
   2783         )
   2784         if isinstance(f, RepoFile)
   2785     ]

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/huggingface_hub/hf_api.py:2915, in HfApi.list_repo_tree(self, repo_id, path_in_repo, recursive, expand, revision, repo_type, token)
   2913 encoded_path_in_repo = "/" + quote(path_in_repo, safe="") if path_in_repo else ""
   2914 tree_url = f"{self.endpoint}/api/{repo_type}s/{repo_id}/tree/{revision}{encoded_path_in_repo}"
-> 2915 for path_info in paginate(path=tree_url, headers=headers, params={"recursive": recursive, "expand": expand}):
   2916     yield (RepoFile(**path_info) if path_info["type"] == "file" else RepoFolder(**path_info))

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/huggingface_hub/utils/_pagination.py:36, in paginate(path, params, headers)
     34 session = get_session()
     35 r = session.get(path, params=params, headers=headers)
---> 36 hf_raise_for_status(r)
     37 yield from r.json()
     39 # Follow pages
     40 # Next link already contains query params

File ~/miniforge3/envs/optimum/lib/python3.11/site-packages/huggingface_hub/utils/_errors.py:352, in hf_raise_for_status(response, endpoint_name)
    333 elif error_code == "RepoNotFound" or (
    334     response.status_code == 401
    335     and response.request is not None
   (...)
    342     # => for now, we process them as `RepoNotFound` anyway.
    343     # See https://gist.github.com/Wauplin/46c27ad266b15998ce56a6603796f0b9
    344     message = (
    345         f"{response.status_code} Client Error."
    346         + "\n\n"
   (...)
    350         " make sure you are authenticated."
    351     )
--> 352     raise RepositoryNotFoundError(message, response) from e
    354 elif response.status_code == 400:
    355     message = (
    356         f"\n\nBad request for {endpoint_name} endpoint:" if endpoint_name is not None else "\n\nBad request:"
    357     )

RepositoryNotFoundError: 401 Client Error. (Request ID: Root=1-65e098fa-52dd095c38eb5594565d025c;31802fed-d380-466b-8515-710ff61d6dce)

Repository Not Found for url: https://huggingface.co/api/models/[username]/[repository]/tree/main?recursive=True&expand=False.
Please make sure you specified the correct `repo_id` and `repo_type`.
If you are trying to access a private or gated repo, make sure you are authenticated.
Invalid username or password.

**Comment by fxmarty** [comment]
@idObTri Thank you for the report, I can not reproduce the issue. Can you try: `pip install -U optimum huggingface_hub transformers`

and
```
huggingface-cli login
```

with a token from https://huggingface.co/settings/tokens?

```python
from optimum.onnxruntime import ORTModelForCustomTasks

model = ORTModelForCustomTasks.from_pretrained("fxmarty/tiny-onnx-private-2")
```

works for me with optimum==1.17.1, huggingface-hub==0.21.3 & transformers==4.38.1

**Comment by idObTri** [comment]
Thank you for the reply.

I updated each module to optimum==1.17.1, huggingface-hub==0.21.3 and transformers==4.39.0dev0, but still had this problem.
Next, following your advice, I used huggingface_hub.login(hf_token) instead of use_auth_token = hf_token with ORTModelForCustomTasks(). This works well.
(with my notebook, huggingface-cli didn't work well so I tried to use huggingface_hub module)

Should I move to migrate to using huggingface-cli / hugginface_hub from use_auth_token?
